src/StarRailService.test.ts

```diff
@@ -0,0 +1,4711 @@
+import { expect, test, describe, beforeAll } from "bun:test";
+
+import MOCK_USER_INFO from "./_mocks/userinfo_v1_3_0.json";
+import MOCK_USER_INFO_A from "./_mocks/userinfo_v1_3_0_a.json";
+import MOCK_USER_INFO_B from "./_mocks/userinfo_v1_3_0_b.json";
+import MOCK_USER_INFO_C from "./_mocks/userinfo_v1_3_0_c.json";
+import { Character, User } from "./game";
+import { StarRailApi } from "./api";
+import { StarRailService } from "./StarRailService.ts";
+import { ApiCharacterData } from "./api/StarRailApi.type.ts";
+
+type ComparisonResult = { success: boolean; message: string };
+
+// Write a custom comparer for nested Objects, since Bun currently doesn't support `objectContaining`
+// > https://github.com/oven-sh/bun/issues/1825#issuecomment-1681785950
+function deepCompare<T>(obj1: T, obj2: T): ComparisonResult {
+  if (obj1 === obj2) return { success: true, message: "All properties match" };
+
+  if (typeof obj1 !== typeof obj2) {
+    return {
+      success: false,
+      message: `Type mismatch: expected ${typeof obj1}, got ${typeof obj2}`,
+    };
+  }
+
+  if (typeof obj1 !== "object" || obj1 === null || obj2 === null) {
+    return {
+      success: false,
+      message: `Value mismatch: expected ${obj1}, got ${obj2}`,
+    };
+  }
+
+  for (const key in obj1) {
+    if (Object.prototype.hasOwnProperty.call(obj1, key)) {
+      if (!Object.prototype.hasOwnProperty.call(obj2, key)) {
+        return { success: false, message: `Missing property: ${key}` };
+      }
+      const result = deepCompare((obj1 as never)[key], (obj2 as never)[key]);
+      if (!result.success) {
+        return {
+          success: false,
+          message: `Property "${key}" does not match: ${result.message}`,
+        };
+      }
+    }
+  }
+
+  return { success: true, message: "All properties match" };
+}
```

Comment:
Wrote a custom matcher because Bun currently doesn't support `objectContaining`.
Related:
- https://github.com/oven-sh/bun/issues/1825#issuecomment-1681785950